### PR TITLE
implement inactive activity indicator on progress bar

### PR DESCRIPTION
**Pull request by fengyun5264**
# feature request

The rrweb-player shows an indicator for snapshot events but it would also be nice, to have some sort of indication of the inactive portion of the recording during playback. Recorded events can span to several minutes and it might be useful to know before hand where the inactive part is so it can be skipped manually as-well

# tasks

- [x] a util function for detecting inactive periods
- [x] implementation of indicator UI on progress bar
- [x] get session events in Control.svelte and invoke util function to get invalid periods
- [x] calculate width of indicator according to the period length
- [x] combine calculation value with indicator UI
- [x]  manual test and bug fix
- [x] improve code quality

### CODE REVIEW

**Comment by HurricaHjz** [comment]
Feature added: a fixed div element can be added as an indicator

**Comment by MengZihan712** [comment]
Feature added: can add a variable inactivePeriods and use util function to get inactive periods.

**Comment by u6924169** [comment]
Feature added: can add width property for inactive activity indicators.

**Comment by DexxDing** [comment]
Feature added: can combine calculation value with indicator UI

File: packages/rrweb-player/src/utils.ts [hunk 1]
```diff
@@ -141,3 +144,31 @@ export function typeOf(
   // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access
   return map[toString.call(obj)];
 }
+
+
+function isUserInteraction(event: eventWithTime): boolean {
```
Comment (DexxDing) [inline packages/rrweb-player/src/utils.ts:149]: IMG_7877
There appear to be a type error in this function

File: packages/rrweb-player/src/utils.ts [hunk 2]
```diff
@@ -141,3 +144,31 @@ export function typeOf(
...
+    event.data.source > IncrementalSource.Mutation &&
+    event.data.source <= IncrementalSource.Input
+  );
+}
+
+
+const SKIP_TIME_THRESHOLD = 10 * 1000;
+
+
+export function getInactivePeriods(events: eventWithTime[]) {
```
Comment (DexxDing) [inline packages/rrweb-player/src/utils.ts:163]: More coments can be added to this function to benifit other contributors

**Review by DexxDing — CHANGES_REQUESTED** [review]

File: packages/rrweb-player/src/Controller.svelte [hunk 3]
```diff
@@ -95,6 +94,37 @@
     return customEvents;
   })();
 
+  let inactivePeriods: {
+    name: string;
+    background: string;
+    position: string;
+    width: string;
+  }[];
+  $: inactivePeriods = (() => {
```
Comment (MengZihan712) [inline packages/rrweb-player/src/Controller.svelte:103]: My suggestion is that putting the whole function in a try-catch code structure would make the code safer.

Comment (fengyun5264) [reply packages/rrweb-player/src/Controller.svelte:103]: @MengZihan712 Thanks, good suggestion! I will add this soon.

**Review by MengZihan712 — APPROVED** [review]

**Review by MengZihan712 — APPROVED** [review]

**Review by DexxDing — APPROVED** [review]

**Review by HurricaHjz — APPROVED** [review]
good work!

**Pull request merged**